### PR TITLE
DENG-774 Change Control for active_users_aggregates

### PR DESCRIPTION
**Pull request by lucia-vargas-a**
Implements [DENG-774](https://mozilla-hub.atlassian.net/browse/DENG-774)

For modifications to schemas in restricted namespaces (see [`CODEOWNERS`](https://github.com/mozilla/bigquery-etl/blob/main/CODEOWNERS)):
- [x] Follow the [change control procedure](https://docs.google.com/document/d/1TTJi4ht7NuzX6BPG_KTr6omaZg70cEpxe9xlpfnHj9k/edit#heading=h.ttegrcfy18ck)

[DENG-774]: https://mozilla-hub.atlassian.net/browse/DENG-774?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: bigquery_etl/cli/query.py
Comment (lucia-vargas-a) [inline bigquery_etl/cli/query.py:1127]: Generating sql only for files with change control implemented.

**Review by bochocki — APPROVED** [review]
This LGTM! I don't think I fully understand how all the pieces fit together with the bigger picture (e.g. I wouldn't have caught @scholtzan's comment about the Airflow task), but from what I'm able to follow it looks great. I just had a few minor refactoring suggestions.

File: bigquery_etl/cli/query.py
Comment (scholtzan) [inline bigquery_etl/cli/query.py:1875]:
```suggestion
    """Validate the defined query schema against the query and destination table."""
```

File: bigquery_etl/cli/query.py
Comment (scholtzan) [inline bigquery_etl/cli/query.py:1839]: I think we want to keep this

File: bigquery_etl/metadata/validate_metadata.py
Comment (scholtzan) [inline bigquery_etl/metadata/validate_metadata.py:54]:
```suggestion
            if "@" not in owner:
```

File: bigquery_etl/metadata/validate_metadata.py
Comment (scholtzan) [inline bigquery_etl/metadata/validate_metadata.py:50]: This could also be a boolean flag `has_codeowner` that gets set if the owner is in the CODEOWNERS file. Could make the condition a little more readable

File: bigquery_etl/metadata/validate_metadata.py
Comment (lucia-vargas-a) [reply bigquery_etl/metadata/validate_metadata.py:42]: Only using it because the default values were not properly retrieved. Let me see what I can do to avoid these lines of code.

File: CODEOWNERS
Comment (lucia-vargas-a) [inline CODEOWNERS:5]: @bochocki preferences about the name of this group? I'll check when it's available before merging the PR.

File: tests/cli/test_cli_metadata.py
Comment (lucia-vargas-a) [reply tests/cli/test_cli_metadata.py:18]: Great Brad, thank you. Implemented with some refactoring of the tests.

File: CODEOWNERS
Comment (bochocki) [reply CODEOWNERS:5]: This sounds good to me!

**Review by whd — COMMENTED** [review]
I had some trouble reading through the validation section. I think it would be simpler to reason about if we required all github identities listed as `owners` in metadata to be listed as CODEOWNERS for tables subject to this change control.

File: CODEOWNERS
Comment (lucia-vargas-a) [reply CODEOWNERS:5]: Resolved.

File: bigquery_etl/metadata/validate_metadata.py
Comment (lucia-vargas-a) [reply bigquery_etl/metadata/validate_metadata.py:49]: Yes, in this case, the team includes people from both teams and that is aligned with the requirement.
Thank you for the link, noted in case the requirement changes in the future.

File: bigquery_etl/metadata/validate_metadata.py
Comment (lucia-vargas-a) [reply bigquery_etl/metadata/validate_metadata.py:70]: Slight change of syntax.

File: dags/bqetl_analytics_aggregations.py
Comment (scholtzan) [inline dags/bqetl_analytics_aggregations.py:121]: Will this cause any issues when sending emails since it is not technically an email address?

**Review by scholtzan — APPROVED** [review]

File: bigquery_etl/metadata/validate_metadata.py
Comment (whd) [inline bigquery_etl/metadata/validate_metadata.py:70]: I'm still somewhat confused about the use of `occurrences` here, which will always be `0` or `1`, since this check appears to require that there be at least one line matching the format `row_to_search_for`. That value is generated for each metadata owner, one at a time, so per the string comparison strictly requires that there exist a `CODEOWNERS` line containing only a single `CODEOWNER` from among those specified in metadata. Therefore I would expect the following to fail the check:

```
/{path_in_codeowners} {meta_owner} {other_meta_owner}
```

which is particularly confusing because  the error message that is emitted presumably looks like:

`The metadata includes label change_controlled, which requires CODEOWNERS to include at least one owner's record: [{meta_owner}, {other_meta_owner}]`

The requirements in the error message are met by my example (`two` relevant CODEOWNERS is at least `one`). The most trivial way to deal with this is to change the wording and specification to require `exactly one owner's record` instead of `at least one owner's record`, to match what the check implements, but I think we should relax the parsing of the CODEOWNERS line to match the natural language specification and error message.

Related but separate, due to [precedence semantics](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners#codeowners-syntax), a CODEOWNERS file might pass the CI check but fail to enforce the expected change control via the following:

```
/{path_in_codeowners} @meta-owner
/{path_in_codeowners} @user-not-in-codeowners
```

This should be caught by a human reviewing the PR so it's perhaps not worth explicitly verifying in CI, but given that the we're implementing these checks to be very specific about change control I'm noting it as an edge case.

File: dags/bqetl_analytics_aggregations.py
Comment (lucia-vargas-a) [reply dags/bqetl_analytics_aggregations.py]: I think it would. I added a change to [remove non-email owners ](https://github.com/mozilla/bigquery-etl/blob/08e997a2d80fd52434cba6584b4a6a64c1fa67fb/bigquery_etl/query_scheduling/task.py#L333-L341) when generating the email list for Airflow.

File: bigquery_etl/metadata/validate_metadata.py
Comment (whd) [inline bigquery_etl/metadata/validate_metadata.py:69]: Since the logic has been refactored back to using substring comparisons, I think the comments from my first review are relevant again about accepting commented strings e.g. I would expect this to erroneously pass the check:

```
# {path_in_codeowners} {valid_owner}
```

Can you add a test for this, as well as a test verifying that from my second review a line like:

```
{path_in_codeowners} {valid_owner_1} {valid_owner_2}
```

is now accepted?

Comment (lucia-vargas-a) [reply bigquery_etl/metadata/validate_metadata.py]: Agree, added a few test scenarios for github identities, one/none/many owners in CODEOWNERS and the handling of lines starting with '#"

File: bigquery_etl/metadata/validate_metadata.py
Comment (whd) [inline bigquery_etl/metadata/validate_metadata.py:61]: I suggest moving the `startswith` check here to avoid duplicating it below.

**Review by whd — APPROVED** [review]
r+wc

Thank you for adding the tests. I think it's important to decide whether to require only `owners` in CODEOWNERS, but I don't think that needs another round of review if a test is added.

After this is merged https://github.com/mozilla/bigquery-etl/settings/branch_protection_rules/4491948 should be updated to require the `validate-metadata` check since it isn't required presently.

File: bigquery_etl/metadata/validate_metadata.py
Comment (lucia-vargas-a) [reply bigquery_etl/metadata/validate_metadata.py:61]: Yes, also to not validate lines unnecessarily.

**Pull request merged**
